src/hardware/compressor.h

```diff
@@ -0,0 +1,100 @@
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+/* This is a simplified port of the "Master Tom Compressor" JSFX effect
+ * bundled with REAPER, originally written by Thomas Scott Stillwell.
+ * Copyright notice of the original effect plugin:
+ *
+ *
+ * Copyright 2006, Thomas Scott Stillwell
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * The name of Thomas Scott Stillwell may not be used to endorse or promote
+ * products derived from this software without specific prior written
+ * permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DOSBOX_COMPRESSOR_H
+#define DOSBOX_COMPRESSOR_H
+
+typedef struct AudioFrame AudioFrame_;
+
+class Compressor {
+public:
+	Compressor();
+	~Compressor();
+
+	void Configure(const uint16_t sample_rate_hz, const float threshold_db,
+	               const float ratio, const float release_time_ms,
+	               const float rms_window_ms);
+	void Reset();
+
+	AudioFrame Process(const AudioFrame &in);
+
+	// prevent copying
+	Compressor(const Compressor &) = delete;
+	// prevent assignment
+	Compressor &operator=(const Compressor &) = delete;
+
+private:
+	uint16_t sample_rate_hz = {};
+
+	static constexpr auto num_attack_times = 120;
+	double attack_times_ms[num_attack_times] = {};
```

Comment:
Maybe an array-type; then can use .size() on it?
``` c++
std::array<float, 120> attack_times_ms = {};
```

Reply:
Ah yeah, I almost did it that way.